core/src/main/java/org/neo4j/ogm/context/EntityGraphMapper.java

```diff
@@ -211,36 +215,23 @@ private void deleteObsoleteRelationships() {
 
                 // remove all nodes that are referenced by this relationship in the mapping context
                 // this will ensure that stale versions of these objects don't exist
-                clearRelatedObjects(mappedRelationship.getStartNodeId());
-                clearRelatedObjects(mappedRelationship.getEndNodeId());
-
-                // finally remove the relationship from the mapping context
-                //mappingContext.removeRelationship(mappedRelationship);
-                mappedRelationshipIterator.remove();
+                staleNodeIds.add(mappedRelationship.getStartNodeId());
+                staleNodeIds.add(mappedRelationship.getEndNodeId());
             }
         }
-    }
-
-    private void clearRelatedObjects(Long node) {
 
-        for (MappedRelationship mappedRelationship : mappingContext.getRelationships()) {
-            if (mappedRelationship.getStartNodeId() == node || mappedRelationship.getEndNodeId() == node) {
-
-                Object dirty = mappingContext.getNodeEntity(mappedRelationship.getEndNodeId());
-                if (dirty != null) {
-                    LOGGER.debug("flushing end node of: (${})-[:{}]->(${})", mappedRelationship.getStartNodeId(),
-                        mappedRelationship.getRelationshipType(), mappedRelationship.getEndNodeId());
-                    mappingContext.removeNodeEntity(dirty, true);
-                }
-
-                dirty = mappingContext.getNodeEntity(mappedRelationship.getStartNodeId());
-                if (dirty != null) {
-                    LOGGER.debug("flushing start node of: (${})-[:{}]->(${})", mappedRelationship.getStartNodeId(),
-                        mappedRelationship.getRelationshipType(), mappedRelationship.getEndNodeId());
-                    mappingContext.removeNodeEntity(dirty, true);
-                }
-            }
-        }
+        do {
+            // Remove all nodes that are not isolated (that is only in one relationship)
+            // Don't do this as a filter, as this list is needed to calculate further stalte nodes
+            staleNodeIds.removeIf(canNotBeStale);
+            // Remove all the stale nodes
+            staleNodeIds.stream()
+                .map(mappingContext::getNodeEntity)
+                .filter(Objects::nonNull)
+                .forEach(node -> mappingContext.removeNodeEntity(node, true));
+            // Remove now stale relationship and possibly creating new stale nodes
+            staleNodeIds = mappingContext.removeStaleRelationships(staleNodeIds);
+        } while (!staleNodeIds.isEmpty());
```

Comment:
Why `do..while` and not `while`? It seems at least possible that the list is empty. I personally prefer in such cases to avoid the function calls and put the loop conditions on top of the loop.